File: src/main/java/com/jloom/scaffold/FileTreeCopier.java
package com.jloom.scaffold;

import com.jloom.registry.CatalogRoots;
import com.jloom.registry.ModuleManifest;
import com.jloom.util.Tokens;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;

public final class FileTreeCopier {

    public void copy(ModuleManifest manifest, Path targetRoot, Map<String, String> tokens) {
        for (String relativePath : manifest.fileTemplates()) {
            String destinationRelativePath = Tokens.substitute(relativePath, tokens);
            Path destination = targetRoot.resolve(destinationRelativePath);

            try (InputStream in = CatalogRoots.open(manifest.id(), "files/" + relativePath)) {
                if (in == null) {
                    throw new IllegalStateException(
                            "Module '" + manifest.id() + "' declares fileTemplate '" + relativePath
                                    + "' but no such resource exists under files/");
                }
                Files.createDirectories(destination.getParent());
                if (isBinary(relativePath)) {
                    Files.write(destination, in.readAllBytes());
                } else {
                    String content = new String(in.readAllBytes(), StandardCharsets.UTF_8);
                    Files.writeString(destination, Tokens.substitute(content, tokens));
                }
                if (isExecutableScript(relativePath)) {
                    destination.toFile().setExecutable(true, false);
                }
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        }
    }

    private static boolean isBinary(String path) {
        return path.endsWith(".jar") || path.endsWith(".png") || path.endsWith(".ico");
    }

    private static boolean isExecutableScript(String path) {
        if (path.endsWith(".bat") || path.endsWith(".cmd")) {
            return false;
        }
        String fileName = Path.of(path).getFileName().toString();
        return fileName.equals("gradlew")
                || fileName.endsWith(".sh")
                || fileName.endsWith(".command");
    }
}

File: node/src/scaffold/fileTreeCopier.ts
import { chmodSync, mkdirSync, writeFileSync } from "node:fs";
import path from "node:path";
import { readBytes, readText, resolvePath } from "../registry/catalogRoots.ts";
import { substitute } from "../merge/tokens.ts";
import type { ModuleManifest } from "../registry/types.ts";

function isBinary(relativePath: string): boolean {
  return relativePath.endsWith(".jar") || relativePath.endsWith(".png") || relativePath.endsWith(".ico");
}

function isExecutableScript(relativePath: string): boolean {
  if (relativePath.endsWith(".bat") || relativePath.endsWith(".cmd")) {
    return false;
  }
  const fileName = path.basename(relativePath);
  return fileName === "gradlew" || fileName.endsWith(".sh") || fileName.endsWith(".command");
}

/** Mirrors FileTreeCopier.copy(...): copies a module's fileTemplates into the target project,
 * substituting {{tokens}} into both the destination path and (for text files) the content.
 * Binary files (.jar/.png/.ico) are copied byte-for-byte with no substitution. */
export function copy(manifest: ModuleManifest, targetRoot: string, tokens: Record<string, string>): void {
  for (const relativePath of manifest.fileTemplates) {
    const destinationRelativePath = substitute(relativePath, tokens);
    const destination = path.join(targetRoot, destinationRelativePath);
    const sourceRelative = `files/${relativePath}`;

    if (resolvePath(manifest.id, sourceRelative) === undefined) {
      throw new Error(
        `Module '${manifest.id}' declares fileTemplate '${relativePath}' but no such resource exists under files/`,
      );
    }

    mkdirSync(path.dirname(destination), { recursive: true });

    if (isBinary(relativePath)) {
      const bytes = readBytes(manifest.id, sourceRelative)!;
      writeFileSync(destination, bytes);
    } else {
      const content = readText(manifest.id, sourceRelative)!;
      writeFileSync(destination, substitute(content, tokens), "utf8");
    }

    if (isExecutableScript(relativePath)) {
      chmodSync(destination, 0o755);
    }
  }
}
